microsoft/surface/kernel/default.nix

```diff
@@ -0,0 +1,471 @@
+          # Surface Hotplug
+          #
+          SURFACE_HOTPLUG m
+
+          #
+          # IPTS touchscreen
+          #
+          # This only enables the user interface for IPTS data.
+          # For the touchscreen to work, you need to install iptsd.
+          #
+          MISC_IPTS m
+
+          #
+          # Cameras: IPU3
+          #
+          ## TODO: Fix for kernel 5.10.4:
+          ##VIDEO_IPU3_IMGU m
+          ##VIDEO_IPU3_CIO2 m
+          ##CIO2_BRIDGE y
+          ##INT3472 m
+
+          #
+          # Cameras: Sensor drivers
+          #
+          ## TODO: Fix for kernel 5.10.4:
+          ##VIDEO_OV5693 m
+          ##VIDEO_OV8865 m
+
+          #
+          # Other Drivers
+          #
+          INPUT_SOC_BUTTON_ARRAY m
+          SURFACE_3_BUTTON m
+          SURFACE_3_POWER_OPREGION m
+          SURFACE_PRO3_BUTTON m
+          SURFACE_GPE m
+          SURFACE_BOOK1_DGPU_SWITCH m
+        '';
+      }
+      {
+        name = "ms-surface/0001-surface3-oemb";
+        patch = patches + "/5.10/0001-surface3-oemb.patch";
+      }
+      ## TODO: Fix for kernel 5.10.4
+      # {
+      #   name = "ms-surface/0002-wifi";
+      #   patch = patches + "/5.10/0002-wifi.patch";
+      # }
+      {
+        name = "ms-surface/0003-ipts";
+        patch = patches + "/5.10/0003-ipts.patch";
+      }
+      {
+        name = "ms-surface/0004-surface-gpe";
+        patch = patches + "/5.10/0004-surface-gpe.patch";
+      }
+      {
+        name = "ms-surface/0005-surface-sam-over-hid";
+        patch = patches + "/5.10/0005-surface-sam-over-hid.patch";
+      }
+      {
+        name = "ms-surface/0006-surface-sam";
+        patch = patches + "/5.10/0006-surface-sam.patch";
+      }
+      {
+        name = "ms-surface/0007-surface-hotplug";
+        patch = patches + "/5.10/0007-surface-hotplug.patch";
+      }
+      {
+        name = "ms-surface/0008-surface-typecover";
+        patch = patches + "/5.10/0008-surface-typecover.patch";
+      }
+      ## TODO: Fix for kernel 5.10.4:
+      # {
+      #   name = "ms-surface/0009-cameras";
+      #   patch = patches + "/5.10/0009-cameras.patch";
+      # }
+    ];
+  };
+
+  surface_kernelPackages = {
+    linux_5_4_24 = (with pkgs; recurseIntoAttrs (linuxPackagesFor (
+      callPackage ./linux-5.4.24.nix {
+        kernelPatches = [
+          kernelPatches.bridge_stp_helper
+          # kernelPatches.request_key_helper
+          # kernelPatches.export_kernel_fpu_functions
+        ]
+        ++ surface_kernelPatches.linux_5_4;
+      }
+    )));
+
+    linux_5_5_8 = (with pkgs; recurseIntoAttrs (linuxPackagesFor (
+      callPackage ./linux-5.5.8.nix {
+        kernelPatches = [
+          kernelPatches.bridge_stp_helper
+          # kernelPatches.request_key_helper
+          # kernelPatches.export_kernel_fpu_functions
+        ]
+        ++ surface_kernelPatches.linux_5_5;
+      }
+    )));
+
+    linux_5_9_2 = (with pkgs; recurseIntoAttrs (linuxPackagesFor (
```

Comment:
Pinning kernel versions seems reasonable. Do you think you can maintain/test three kernel versions?

Reply:
I'm not keen to support that many kernels, but left them in for the purpose of the draft — to get some feedback in case I'm going the wrong direction:
- 20.09 doesn't seem to support 5.10, yet.
- After 5.4 the ipts driver changed; I'm working on getting `iptsd` built from @hpfr's work.

Reply:
Actually, is it acceptable for me to raise multiple smaller PRs, instead?
I normally prefer that, to simplify review, therefore thinking that I should perhaps leave this PR for the kernel patching alone, and raise new PRs for the other work, for example the `ipstd` touch-screen daemon.

Reply:
I don't mind reviewing this or smaller PRs, right now the PR is not too big for me. I would suggest however to reduce the number of kernel to what you actually test. Keeping one kernel up-to-date is tedious enough usually.

Reply:
Thanks.
I'm already planning to drop the extra kernels.
The code strikes me as a little messy and hard to follow with them in.

There will def have to be a follow-up PR to deal with some problems I've discovered, and am working through, with the nonfree linux firmware from kernel.org;  it's arguably independent of this PR, though.

I've been gathering notes to add to a README.